### PR TITLE
geotile_grid implementation

### PR DESCRIPTION
**Pull request by nyurik**
Implements `geotile_grid` aggregation, refactoring previous implementation #30240

This code uses the same base classes as `geohash_grid` agg, but uses a different hashing
algorithm to allow zoom consistency.  Each grid bucket is aligned to Web Mercator tiles.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java]: Tiny nit, should we put parenthesis around the components here just so the order of operation is immediately clear?

And not knowing anything about the algo, is there a chance `tiles` could ever be zero (e.g. divide by zero scenario)?

File: server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java]: Style nit: I think we try to avoid conditionals without braces, too easy to accidentally introduce bugs if you're not paying attention when touching the code.

Could replace with `xTile = Math.min(xTile, 0)`, if you're looking for a concise alternative.  Probably similar for the rest too.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/InternalQuadkeyGrid.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/InternalQuadkeyGrid.java]: Think this name needs to change?  And ditto for javadoc on the class

File: server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java]: Two questions:

1. Should we support unit-less precision levels?  It looks like it adds a lot of autodetect magic below, and we've consistently been moving towards requiring units everywhere in ES.  I don't know enough about the algo to say, but is there something tangible that we gain by supporting unit-less precision over just `10km`, etc?
2. Is there a reason we fall back to the "old" style of xcontent parsing instead of using the newer static parsers here too?

**Review by polyfractal — COMMENTED** [review]
Sorry, I know this is labeled WIP.  I was just passing by and noticed something I wanted to leave a comment on, then left a few more.  Feel free to ignore if you're still iterating on things :)

File: server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java
Comment (nyurik) [reply server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java]: 1) for quadkey, most of the time it will be unitless - e.g .map visualizations will use `current_zoom + 2` as the precision to get 16 dots per visualization tile.
2) I tried to make as few changes as possible to the existing code to make reviewing easier. This code is an exact copy of the `GeoUtils.parsePrecision()`, adjusted for different validation and parsing. We might consider rewriting them, but I think we should do them at the same time and possibly consolidate it for reuse.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/InternalQuadkeyGrid.java
Comment (nyurik) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/InternalQuadkeyGrid.java]: thx, good catch. The NAME shouldn't have even been here - its a dup. FIxed here and in InternalGeoHashGrid.

File: server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java
Comment (nyurik) [reply server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java]: thx, refactored.

File: server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java
Comment (nyurik) [reply server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java]: I refactored to put all logic in a single method. Tiles is a result of (4^N), so cannot be 0

File: server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java
Comment (nyurik) [reply server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java]: @polyfractal, per @talevy 's suggestion, I removed non-integer precision - it is not needed by our use cases, and we do not know if they will be needed by other users. This also simplifies precision parsing.

File: server/src/main/java/org/elasticsearch/search/aggregations/AggregationBuilders.java
Comment (nyurik) [reply server/src/main/java/org/elasticsearch/search/aggregations/AggregationBuilders.java]: @imotov I added more tests to use this function. Let me know if anything else is needed. Also, there were some unrelated CI issues, hopefully master merge will fix them.

**Review by imotov — APPROVED** [review]
LGTM. But since I didn't sleep for the last 27 hours, it might make sense for @polyfractal to have another look.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileUtils.java
Comment (talevy) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileUtils.java]: are we concerned that the latitude is not clipped here to ensure that the projection is square?

```
latitude = Math.min(Math.max(lat,  -85.05112878),  85.05112878);
```

this is regarding:

> The latitude and longitude are assumed to be on the WGS 84 datum. Even though Bing Maps uses a spherical projection, it’s important to convert all geographic coordinates into a common datum, and WGS 84 was chosen to be that datum. The longitude is assumed to range from -180 to +180 degrees, and the latitude must be clipped to range from -85.05112878 to 85.05112878. This avoids a singularity at the poles, and it causes the projected map to be square. [1]

and

> Because the Mercator projects the poles at infinity, Google Maps cannot show the poles. Instead it cuts off coverage at 85.051129° north and south. This is not considered a limitation, given the purpose of the service. The value 85.051129° is the latitude at which the full map becomes a square...[2]

[1] https://docs.microsoft.com/en-us/bingmaps/articles/bing-maps-tile-system
[2] https://en.wikipedia.org/wiki/Web_Mercator_projection

Comment (talevy) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileUtils.java]: also, should there be tests for points in this region?

**Review by talevy — CHANGES_REQUESTED** [review]
Overall looks good! left a few questions

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileUtils.java
Comment (nyurik) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileUtils.java]: We already test for points in that region, e.g.  `assertEquals(0x77FFFF4580000000L, longEncode(179.999, 89.999, 29));`, but i will add a few more. The mathematics automatically takes care of those values, assigning geopoints in that region with the min/max Y value.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileUtils.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileUtils.java]: Should we just make all the relevant variables/static defaults `long` instead of `int`, to save on all the casting?  Not because it matters for performance, but just for readability.

I don't think we're relying on int overflows anywhere, so it should be safe to just use longs?

I don't care strongly, so feel free to ignore this (or if I missed somewhere else that needs them as ints).  Was just something I noticed in passing :)

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileUtils.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileUtils.java]: Just curious about this arrangement... what's the reason for just catching `IllegalArgumentException` and then rethrowing it?  Should we just catch any `Exception`?  I'm assuming this is here to make the exception message a little nicer?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/ParsedGeoTileGridBucket.java
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/ParsedGeoTileGridBucket.java]: Can the hash be null? (e.g. curious about `p.textOrNull()`.  Won't that cause `keyToGeoPoint()` to NPE later?

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileGridAggregatorTests.java
Comment (polyfractal) [inline server/src/test/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileGridAggregatorTests.java]: Could we add a few tests for invalid construction (`expectThrows()` on bad precision, bad zoom, etc) to this suite?  I know it's indirectly tested in `GeoTileGridParserTests`, but that's done through JSON parsing and not the builder itself.

:heart:

**Review by polyfractal — APPROVED** [review]
Left a few minor questions/comments, but LGTM otherwise.  Will defer to @talevy on algo review :)

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileGridAggregatorTests.java
Comment (nyurik) [reply server/src/test/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileGridAggregatorTests.java]: I added an extra test to `GeoTileGridAggregatorTests` to test precision. Also, added a few more tests to the utils class to check for invalid input vals.

**Review by talevy — APPROVED** [review]
LGTM if CI is happy.

thanks for all the iterations Yuri, I like where it landed!

**Pull request merged**
